File: 每日一题/5.13/3335.字符串转换后的长度-i.cpp
/*
 * @lc app=leetcode.cn id=3335 lang=cpp
 *
 * [3335] 字符串转换后的长度 I
 */
#include "tools.h"
// @lc code=start
class Solution {
public:
    const int MOD = 1'000'000'007;
    int lengthAfterTransformations(string s, int t) {
        vector<int> cur(26, 0);
        for (auto ch : s)
        {
            cur[ch - 'a'] += 1;
        }

        while (t --)
        {
            vector<int> nxt(26, 0);
            nxt[0] = cur[25];
            nxt[1] = (cur[25] + cur[0]) % MOD;
            for (int i = 2; i < 26; i ++ ) nxt[i] = cur[i - 1];
            cur = nxt;
        }
        long long res = 0;
        for (int i = 0; i < 26; i ++ )
        {
            res = (res + cur[i]) % MOD;
        }
        return res;
    }
};
// @lc code=end

File: 每日一题/5.13/3335.字符串转换后的长度-i.py
#
# @lc app=leetcode.cn id=3335 lang=python3
#
# [3335] 字符串转换后的长度 I
#
from mytools import *
# @lc code=start
class Solution:
    def lengthAfterTransformations(self, s: str, t: int) -> int:
        cur = [0] * 26
        MOD = 1_000_000_007
        
        for ch in s:
            cur[ord(ch) - ord('a')] += 1
        
        for _ in range(t):
            nxt_state = [0] * 26
            nxt_state[0] = cur[25]
            nxt_state[1] = (cur[0] + cur[25]) % MOD
            for i in range(2, 26):
                nxt_state[i] = cur[i - 1]
            cur = nxt_state
        return sum(cur) % MOD
# @lc code=end

